### PR TITLE
Support knowledge distillation in FastSpeech training

### PR DESCRIPTION
**Pull request by kan-bayashi**
Re-visit #1287 

TODO:
- [x] Add focus rate filtering
- [x] Simplify `run.sh`
- [x] Fix additional input naming in data loader
- [x] Train FastSpeech with Tacotron2 + att-constraint + filtering

### CODE REVIEW

**Comment by kan-bayashi** [comment]
I finished the implementation part.
Now time to review.

**Comment by kan-bayashi** [comment]
And I'm wondering the naming of `expdir`.
I think it is better to add `teacher` model info in the name to be unique but it seems a little bit long.

**Comment by r9y9** [comment]
I’ll take a look tomorrow. Please hold on.

File: egs/ljspeech/tts1/run.sh [hunk 1]
```diff
@@ -38,6 +38,12 @@ train_config=conf/train_pytorch_tacotron2.yaml # you can select from conf or con
                                                # see more info in the header of each config.
 decode_config=conf/decode.yaml
 
+# knowledge distillation related
+teacher_model_path=""
+teacher_decode_config=conf/decode_for_knowledge_dist.yaml
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs/ljspeech/tts1/run.sh]: Sorry I forgot to add. Added.

File: egs/ljspeech/tts1/local/setup_knowledge_dist.sh [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        # shellcheck disable=SC2154
+        ${train_cmd} JOB=1:${nj} "${outdir}/${name}/log/decode.JOB.log" \
+            tts_decode.py \
+                --backend ${backend} \
+                --ngpu 0 \
+                --verbose ${verbose} \
+                --out "${outdir}/${name}/feats.JOB" \
+                --json "${outdir}/${name}/split${nj}utt/data.JOB.json" \
+                --model ${teacher_model_path} \
+                --config ${decode_config}
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs/ljspeech/tts1/local/setup_knowledge_dist.sh]: Those are included in `decode_for_knowledge_dist.yaml`.

File: egs/ljspeech/tts1/local/setup_knowledge_dist.sh [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    for name in ${dev_set} ${train_set}; do
+    (
+        [ ! -e "${outdir}/${name}" ] && mkdir -p "${outdir}/${name}"
+        cp ${dumpdir}/${name}/data.json "${outdir}/${name}"
+        splitjson.py --parts ${nj} "${outdir}/${name}/data.json"
+        # shellcheck disable=SC2154
+        ${train_cmd} JOB=1:${nj} "${outdir}/${name}/log/decode.JOB.log" \
+            tts_decode.py \
+                --backend ${backend} \
+                --ngpu 0 \
```
[the remark this answers is not in the document]
Comment (kan-bayashi) [reply egs/ljspeech/tts1/local/setup_knowledge_dist.sh]: I added npu option.
But keep cpu decoding as default because in the taco2 case it requires around 30 minutes with `nj=32`. That is acceptable, I think.
(But Transformer is too slow. Maybe even if we use GPU.)

File: egs/ljspeech/tts1/local/setup_knowledge_dist.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (r9y9) [reply egs/ljspeech/tts1/local/setup_knowledge_dist.sh]: I see, thanks!

**Comment by kan-bayashi** [comment]
I finished the FastSpeech training with Tacotron.v3 + att-constraint + knowledge distillation.
Here is the sample.
https://drive.google.com/open?id=1peevNormNPaseTMY_Q9yXDITecPt3Fax
It achieves almost the same quality as the teacher model but sometimes lacks the naturalness especially in low- or high-frequency parts.
(e.g. LJ050-0032 or LJ050-0034)

**Comment by kan-bayashi** [comment]
I am also training with tacotron.v2 and transformer.v3 as the teacher.
Please wait for a while (Maybe it will be finished within 2 days).

**Comment by r9y9** [comment]
FastSpeech samples sound great! 

BTW, did you try training with focus rate filtering? I think it's worth trying.

**Comment by kan-bayashi** [comment]
For Transformer, I used focus rate filtering.
It will be finished by the end of today.
Please keep wait for a while!

File: egs/ljspeech/tts1/run.sh [hunk 4]
```diff
@@ -150,8 +157,30 @@ expdir=exp/${expname}
...
-    dt_json=${feat_dt_dir}/data.json
+    if [ -n "${teacher_model_path}" ] && echo "${train_config}" | grep -q "fastspeech"; then
+        # setup feature and duration for fastspeech knowledge distillation training
+        teacher_expdir=$(dirname "$(dirname "${teacher_model_path}")")
+        teacher_outdir=outputs_$(basename ${teacher_model_path})_$(basename ${teacher_decode_config%.*})
+        teacher_outdir=${teacher_expdir}/${teacher_outdir}
+        if [ ! -e ${teacher_outdir}/.done ]; then
+            local/setup_knowledge_dist.sh \
+                --nj ${nj} \
+                --vernose ${verbose} \
```
Comment (r9y9) [inline egs/ljspeech/tts1/run.sh]:
```suggestion
                --verbose ${verbose} \
```

Just found typo.

**Comment by kan-bayashi** [comment]
I added Transformer.v3 + FastSpeech.v4 + focus-rate filtering (thres: 0.9).

**Comment by kan-bayashi** [comment]
Now ready to merge!

**Comment by r9y9** [comment]
Great! How about running ASR objective evaluations? :)

My quick impression is that Tacotron2 is better than Transformer as a teacher model. Possibly alignment failures in our Transformer (which we can confirm in CER/WER)  caused lower quality of training data.

**Comment by kan-bayashi** [comment]
I have ASR results but it is better to add after #1416.
Here are the results.

```
write a CER result in exp/phn_train_no_dev_pytorch_train_tacotron2.v2_fastspeech.v4.single/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.txt
|       Sum/Avg       |        250               21532         |       98.4                 0.4                1.2                 0.4                 2.0               48.0        |
write a WER result in exp/phn_train_no_dev_pytorch_train_tacotron2.v2_fastspeech.v4.single/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.wrd.txt
|       Sum/Avg       |        250                  4023        |       94.8                 3.7                 1.5                  1.6                 6.9                54.0        |

write a CER result in exp/phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.txt
|       Sum/Avg       |        250               21532         |       98.4                 0.4                1.2                 0.3                 2.0               46.4        |
write a WER result in exp/phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.wrd.txt
|       Sum/Avg       |        250                  4023        |       95.0                 3.4                 1.6                  1.6                 6.6                53.2        |

write a CER result in exp/phn_train_no_dev_pytorch_train_transformer.v3_fastspeech.v4.single_filter_fr_thres0.9/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.txt
|        Sum/Avg        |         250                  21532          |        98.2                   0.6                   1.2                   0.6                   2.4                  47.2         |
write a WER result in exp/phn_train_no_dev_pytorch_train_transformer.v3_fastspeech.v4.single_filter_fr_thres0.9/outputs_model.last1.avg.best_decode_denorm.ob_eval/librispeech.transformer.ngpu4_asr.result.v2/phn_eval/result.wrd.txt
|         Sum/Avg        |         250                    4023         |        94.6                    4.0                   1.4                    2.1                   7.5                   54.4         |

```

> My quick impression is that Tacotron2 is better than Transformer as a teacher model. Possibly alignment failures in our Transformer (which we can confirm in CER/WER) caused lower quality of training data. 

I agree with Taco2-based > Transformer-based.
Sometimes, Transformer teacher model contains strange pronunciations (e.g. LJ050-0032).

Also, the loss of Taco2-based is lower than that of Transformer-based.

**with taco2.v3 as a teacher model**
スクリーンショット 2019-12-12 午前0 48 37

**with transformer.v3 as a teacher model**
スクリーンショット 2019-12-12 午前0 49 06

**Comment by r9y9** [comment]
I see you are right. Thank you for providing the complimentary. There looks like a big gap! 

LGTM. Good to merge?
